src/translator/html.cpp

```diff
@@ -0,0 +1,532 @@
+  // NOTE: this should match the exact order of Apply()
+  for (size_t sentenceIdx = 0; sentenceIdx < response.target.numSentences(); ++sentenceIdx) {
+    token_tags_target.push_back(token_tags[token_offset]);  // token_tag for sentence ending gap
+    for (size_t t = 0; t < response.target.numWords(sentenceIdx); ++t) {
+      size_t s = alignments[sentenceIdx][t];
+      assert(s < response.source.numWords(sentenceIdx));
+      token_tags_target.push_back(token_tags[token_offset + 1 + s]);  // +1 for prefix gap
+    }
+
+    token_offset += response.source.numWords(sentenceIdx) + 1;  // +1 for prefix gap
+  }
+
+  assert(token_offset < token_tags.size());
+  token_tags_target.push_back(token_tags[token_offset]);  // token_tag for ending whitespace
+}
+
+AnnotatedText RestoreSource(AnnotatedText const &in, std::vector<HTML::Taint> &token_tags,
+                            std::vector<HTML::Span>::const_iterator span_it,
+                            std::vector<HTML::Span>::const_iterator span_end) {
+  auto prev_it = span_it;  // safe because first span is always empty span, and
+                           // and the while-loop below will do the rest
+
+  // workspace variables for lambda
+  std::string html;
+  HTML::Taint opening, closing;
+
+  return Apply(in, [&](ByteRange range, string_view token, bool last) {
+    // Do encoding of any entities that popped up in the translation
+    // (Also effectively clears html from previous call)
+    EncodeEntities(token, html);
+
+    size_t offset = 0;  // Size added by prepending HTML
+    size_t whitespace_size = CountPrefixWhitespaces(token);
+
+    // Potential issue: spans and tokens can intersect, e.g.
+    //
+    //    text  <p> h <u> e </u> ll o </p>
+    //   spans     |1|   |2|    |3333| (so only 2 is tainted with <p><u>, others only <p>)
+    //  tokens     |111111111111111|2|
+    //
+    // Now 1 covers span 1 to 3, so what taint should it get? Just <p>, or <p><u>?
+
+    // Seek to the last span that overlaps with this token
+    while (true) {
+      DiffTags(prev_it->tags, span_it->tags, opening, closing);
+      prev_it = span_it;
+
+      for (auto cit = closing.crbegin(); cit != closing.crend(); ++cit) {
+        std::string close_tag = format("</{}>", (*cit)->name);
+        html.insert(offset, close_tag);
+        offset += close_tag.size();
+      }
+
+      for (HTML::Tag const *tag : opening) {
+        std::string open_tag = format("<{}{}>", tag->name, tag->attributes);
+        html.insert(offset + whitespace_size, open_tag);
+        offset += open_tag.size();
+      }
+
+      if (span_it + 1 != span_end && ((span_it + 1)->begin < range.end || last)) {
+        span_it++;
+        continue;
+      }
+
+      break;
+    }
+
+    // TODO: This is just the taint of the last span, not the ones in between
+    // I don't know if that is okay for transferring taints. We'll need to test.
+    token_tags.push_back(prev_it->tags);
```

Comment:
This is what makes the empty `<u>` in this case also completely disappear in the reconstructed translation (i.e. `response.target`).